Clean up Tracing API spec, clarify Tracer vs TracerProvider. (#619)
open-telemetry/opentelemetry-specification: specification/trace/api.md

</details>

> Tracing API consist of these main classes:

Reviewer: ```suggestion
The Tracing API consist of these main classes:
```

* The minimal precision is milliseconds.
* The maximal precision is nanoseconds.

## TracerProvider

New `Tracer` instances can be created via a `TracerProvider`.

In implementations of the API, the `TracerProvider` is expected to be the
stateful object that holds any configuration
(such as SpanProcessors and Exporters in the default SDK).
Implementations might require the user to specify such configuration properties
when creating the `TracerProvider`, or rely on external configuration, e.g. when
using the provider pattern.

Normally the `TracerProvider` is expected to be accessed as a singleton, i.e.,

> the API SHOULD provide a single global default `TracerProvider`.

Author: Its described below in the subsection about runtimes. Should I move it back up? Also, I should probably remove the word `single`.

> the API SHOULD provide a single global default `TracerProvider`.

Reviewer 2: I felt like this was missing when I read through this section. You could add a link "See Runtimes for use cases where multiple TracerProviders would be used.".
Well the API itself SHOULD (MUST?) provide one single, global default TracerProvider, right?

> the API SHOULD provide a single global default `TracerProvider`.

Author: I added the link and removed the `single` (I think it is clear that only one global TracerProvider can exist anyway, over-emphasis might be confusing here).

* The minimal precision is milliseconds.
* The maximal precision is nanoseconds.

## TracerProvider

`Tracer`s can be accessed with a `TracerProvider`.

In implementations of the API, the `TracerProvider` is expected to be the
stateful object that holds any configuration
(such as SpanProcessors and Exporters in the default SDK).
Implementations might require the user to specify such configuration properties
when creating the `TracerProvider`, or rely on external configuration, e.g. when

> using the provider pattern.

Reviewer 3: what is "provider pattern"?

Author: Actually I'm not sure, I just move this around. What about removing that "e.g."? Or maybe Java Service Provider Interface (SPI) was meant?

Author: I removed that sentence in 120743dd52163b8c3929740e7bd7f5ff5f6b7b43 as it does not seem to add any value (more confusion).

> Normally the `TracerProvider` is expected to be accessed as a singleton, i.e.,

Reviewer: ```suggestion
Normally, the `TracerProvider` is expected to be accessed as a singleton, i.e.,
```

* The minimal precision is milliseconds.
* The maximal precision is nanoseconds.

## TracerProvider

`Tracer`s can be accessed with a `TracerProvider`.

In implementations of the API, the `TracerProvider` is expected to be the
stateful object that holds any configuration.

> Normally, the `TracerProvider` is expected to be accessed from a central place.

Reviewer 3: Can we say the exact opposite? I am not winning the battle against globals, but I am going to continue fighting it. It's not "normal" to access a global, it's a provably bad design pattern. I would like OpenTelemetry specs to take a stance that "fall back to globals only when nothing else works", not as the first option.

Author: I feel you, but I don't think this PR is the right place for such design decisions.

Reviewer 3: By the same logic, this PR is not a place to introduce this kind of language if it wasn't in the spec previously.

Author: It isn't the right place for that and it was not my intention to strengthen globals in the spec. The current wording even talks about "singletons", which are a design (anti-)pattern to prevent multiple instance of the same object, so I'd say I weakened the wording. See https://github.com/open-telemetry/opentelemetry-specification/blob/v0.5.0/specification/trace/api.md#obtaining-a-tracer

> `TracerProvider`s are generally expected to be used as singletons. Implementations SHOULD provide a single global default `TracerProvider`.

Reviewer 3: I don't consider singleton an anti-pattern - it simply means that there should be only one (*) instance of provider, which is true in most cases. Singleton != global, singletons can be managed properly as dependencies.

(*) as we discussed, the wording about situations where one needs multiple providers is pretty odd. E.g. in a JVM-based app server that runs multiple web apps (like Tomcat), absolutely nothing changes about having a _single provider_ per web app, because class loaders perform the required isolation.

Author: What is called singleton does imply global access. Sure there are other ways to ensure only one instance is created without also providing global access, but that's not the singleton design pattern anymore then.

Thus, the API SHOULD provide a way to set/register and access
a global default `TracerProvider`.

> However, some applications may want to or have to use multiple `TracerProvider`

Reviewer 3: > However, ...

Multiple providers are not the opposite of using a global. Not using a global is the opposite. Multiple providers is just an edge case.

Author: Hmm, I guess the "However" made more sense when the spec still talked about a "singleton" tracer provider (also, "however" seems to be one of my  pet words 😅). Improved in bbc1771625c4fc16e12e2e0489a4c758aef0a7aa, also changed last instance of "settings" to "configuration" like elsewhere.

- `version` (optional): Specifies the version of the instrumentation library
  (e.g. `semver:1.0.0`).

It is unspecified whether or under which conditions the same or different
`Tracer` instances are returned from this functions.

Implementations of this function MUST be prepared that an user calls the
function only once per name+version combination (i.e. caches and re-uses the
returned `Tracer` instance). If configuration changes should be picked up by
users, these changes SHOULD be reflected in already returned objects, not only
if the user calls the API to get a `Tracer` again.

Note: This could, for example, be implemented by
storing any mutable configuration in the `TracerProvider` and
having `Tracer` implementation objects have a reference to the `TracerProvider`
that created them. If configuration must be stored per-tracer
(such as disabling a certain tracer), the tracer could, for example, do a

> look-up with its name+version in a map in the `TracerProvider`, or the

Author: I did not mean to add such level of detail. IMHO, what you describe here is one (good) way to implement what you suggested (the getConfigurationFor method would probably be package-private).

> look-up with its name+version in a map in the `TracerProvider`, or the

Reviewer 2: Yeah the spec should not dictate such implementation details anyway but I was a bit surprised that having the Tracer look it up in the TracerProvider's map is listed as the first example. Since it's just a suggestion it doesn't really matter anyway, however.

> look-up with its name+version in a map in the `TracerProvider`, or the

Author: That's because it allows only storing in the map Tracers that have a non-default configuration and keeps Tracers lightweight. But yes, these sentences are non-normative.

`TracerProvider` could maintain a registry of all returned `Tracer`s and
actively update their configuration if it changes.

### Runtimes with multiple deployments/applications

Some applications may have to use multiple `TracerProvider` instances, e.g.
to have different settings (e.g. `SpanProcessor`s) to each (and
consequently to the `Tracer`s crated by them).

For example, runtimes that support multiple deployments or applications might need to
provide a different `TracerProvider` instance to each deployment. To support this,
a global `TracerProvider` registry may delegate calls to create new instances of
`TracerProvider` to a separate `Provider` component, and the runtime may include
its own `Provider` implementation which returns a different `TracerProvider` for
each deployment.

`Provider` instances are registered with the API via some language-specific
mechanism, for instance the `ServiceLoader` class in Java.

## Tracer

The tracer is responsible for creating `Span`s.

Note: Apart from the library name and version
provided to the `TracerProvider` to get them,

> `Tracer`s are not usually expected to not have mutable state on their own

Author: Better now?

(but they may contain a reference to shared mutable configuration that is
expected to be owned by the TracerProvider).

### Tracer operations

The `Tracer` MUST provide functions to:

- [Create a new `Span`](#span-creation) (see the section on `Span`)

The `Tracer` SHOULD provide methods to:

- Get the currently active `Span`
- Make a given `Span` as active

> The `Tracer` MUST use the [`Context`](../context/Context.md)

Reviewer 2: ```suggestion
The `Tracer` MUST use the [`Context`](../context/context.md)
```
404 :-)

in order to get and set the current `Span` state and to decide how `Span`s
are passed across process boundaries.

When getting the current `Span` while there is none that is currently active,
the `Tracer` MUST return a placeholder `Span` with an invalid `SpanContext`
and `IsRecording` being false.

For any two `Tracer`s returned from the same `TracerProvider`
and queried at the same time in the same logical execution unit

> (see [Context](../context/Context.md)), the active Span

Reviewer 2: ```suggestion
(see [Context](../context/context.md)), the active Span
```

`Tracer` instances are returned from this functions.

Implementations MUST NOT require users to repeatedly obtain a `Tracer` again
with the same name+version to pick up configuration changes.
This can be achieved either by allowing to work with an outdated configuration or
by ensuring that new configuration applies also to previously returned `Tracer`s.

Note: This could, for example, be implemented by storing any mutable
configuration in the `TracerProvider` and having `Tracer` implementation objects
have a reference to the `TracerProvider` from which they were obtained.
If configuration must be stored per-tracer (such as disabling a certain tracer),
the tracer could, for example, do a look-up with its name+version in a map in
the `TracerProvider`, or the `TracerProvider` could maintain a registry of all
returned `Tracer`s and actively update their configuration if it changes.

### Runtimes with multiple deployments/applications <a name="multi-tracerprovider"></a>

Some applications may have to use multiple `TracerProvider` instances, e.g.
to have different settings (like `SpanProcessor`s) for each (and
consequently to the `Tracer`s obtained from them).

> For example, runtimes that support multiple deployments or applications might need to

Reviewer 3: > runtimes that support multiple deployments or applications

such as ...?

Author: I don't know use cases myself although I'm sure there are some. What would you think about removing this whole subsection?

Reviewer: This phrase about "... runtimes that support multiple deployment ..." has been in the spec since very early days (see: 
https://github.com/open-telemetry/opentelemetry-specification/blob/0.1/specification/tracing-api.md#runtime-with-multiple-deploymentsapplications) and only been slightly changed over time.
Not sure what to make it either ...

Author: Then I'd say, let's remove that section (I'll consider the first paragraph separately, that one I moved from the main section).

Author: Removed in fcce2db592bdf246c4fc6b869687625271acd9af, see https://github.com/open-telemetry/opentelemetry-specification/pull/619#discussion_r433762515

provide a different `TracerProvider` instance to each deployment. To support this,

> a global `TracerProvider` registry may delegate calls to create new instances of

Reviewer 3: what is "global `TracerProvider` registry" and why is it even relevant to this section?

The text that follows, esp. "Provider component", is highly confusing.

Author: @z1c0 Can you shed some light on this? Is this still needed?

Reviewer: If I remember correctly, this describes the case of a `TracerFactoryFactory` (note `TracerProvider` was `TracerFactory` originally) for frameworks that rely on a dependency injection mechanism.
Specifically, this was requested by the .NET SIG -> see e.g.  https://github.com/open-telemetry/opentelemetry-dotnet/blob/master/src/OpenTelemetry.Instrumentation.AspNet.Win/README.md

Author: @open-telemetry/dotnet-approvers WDYT?

Author: Since no one objected, I removed that section in fcce2db592bdf246c4fc6b869687625271acd9af

`TracerProvider` to a separate `Provider` component, and the runtime may include
its own `Provider` implementation which returns a different `TracerProvider` for
each deployment.

`Provider` instances are registered with the API via some language-specific
mechanism, for instance the `ServiceLoader` class in Java.

## Tracer

The tracer is responsible for creating `Span`s.

Note: Apart from the library name and version

> provided to the `TracerProvider` to get them,

Reviewer 3: ```suggestion
provided to the `TracerProvider` to get them,
```
```suggestion
passed to the `TracerProvider` to get them,
```

`Tracer`s are not usually expected to not have mutable state of their own
(but they may contain a reference to shared mutable configuration that is

> expected to be owned by the `TracerProvider`).

Reviewer 3: * "library name and version" are not "mutable state", why are they called out?
* "not expected to not have" - double negative
* parentheses are unnecessary
* The value of this note is quite dubious, it speaks about some low-level implementation details that are irrelevant to the API user and only distracts from the main content of the section

Author: I removed the "mutable" from "mutable state" and reworded the sentence to not use parentheses. Also, the double negative was unintentional and wrong, I simplified the sentence in 1e1dd1e8621f8ea1ac174c9d7fdc3c2644aa78fb.

About the note: I do think this note is valuable because it describes how a valid implementation may look like. I think having an concrete example implementation in mind while reading the spec makes the otherwise quite abstract concepts of "TracerProvider" and "Tracer" more easily digestable.

### Tracer operations

The `Tracer` MUST provide functions to:

- [Create a new `Span`](#span-creation) (see the section on `Span`)

The `Tracer` SHOULD provide methods to:

- Get the currently active `Span`
- Make a given `Span` as active

The `Tracer` MUST use the [`Context`](../context/context.md)

> in order to get and set the current `Span` state and to decide how `Span`s

Reviewer 3: ```suggestion
in order to get and set the current `Span` and to decide how `Span`s
```

> are passed across process boundaries.

Reviewer 3: > decide how `Span`s are passed across process boundaries.

is not the responsibility of the tracer, but of the Propagator.

Author: Actually, that's exactly what this note is saying, isn't it? Also, this was not changed in this PR.

When getting the current `Span` while there is none that is currently active,
the `Tracer` MUST return a placeholder `Span` with an invalid `SpanContext`

> and `IsRecording` being false.

Reviewer 3: I don't understand this guideline. I raised this exact issue when tracer names were proposed, and never got a satisfactory implementation proposal of what it means to "disable a tracer". If one were to follow the pattern suggested here, one would end up with orphaned sub-tree in the trace when a normal child span is created from this placeholder span.

Author: You commented on the wrong sentence here, I think. This sentence is about the case where there actually is no active span, so there is no "orphaned sub-tree" here, just a normal root-span.

Author: But I don't want to dodge your original question, see https://github.com/open-telemetry/opentelemetry-specification/pull/619/files#r430566671 (a bit below in the files view)

The `Tracer` MUST internally leverage the `Context` in order to get and set the
current `Span` state and how `Span`s are passed across process boundaries.
The `Tracer` MUST use the [`Context`](../context/context.md)
in order to get and set the current `Span` state and to decide how `Span`s
are passed across process boundaries.

When getting the current span, the `Tracer` MUST return a placeholder `Span`
with an invalid `SpanContext` if there is no currently active `Span`.
When getting the current `Span` while there is none that is currently active,
the `Tracer` MUST return a placeholder `Span` with an invalid `SpanContext`
and `IsRecording` being false.

When creating a new `Span`, the `Tracer` MUST allow the caller to specify the
new `Span`'s parent in the form of a `Span` or `SpanContext`. The `Tracer`

> SHOULD create each new `Span` as a child of its active `Span` unless an

Reviewer 3: you lost this part?

Author: No, this moved to the span creation section where a redundant paragraph already existed.

> returns the placeholder span (see above).

Reviewer 3: Don't understand it. Retrieving an active span from context is not even a function of a tracer, it's a utility function. How can it produce different results?

Author: > Retrieving an active span from context is not even a function of a tracer, it's a utility function.

The spec (before this PR and after) says something different however. Do you think this should be changed to explicitly define that it must return the same as the equivalent Context operation?

> How can it produce different results?

By implementing it like `return isDisabled() ? NullSpan.getInstance() : ContextUtils.currentSpan(Context.current())`.

Author: I can remove this section for now. Probably if we introduce per-tracer disabling, then some explicit API isDisabled would be better. Could also have synergies with #530.

Author: Removed in 60b19a767fd64bcac3d12b1e40ea7868b407e09b

> `Span` SHOULD be made active. A `Span` maybe finished (i.e. have a non-null end

Reviewer 3: ```suggestion
`Span` SHOULD be made active. A `Span` may be finished (i.e. have a non-null end
```

- `version` (optional): Specifies the version of the instrumentation library
  (e.g. `semver:1.0.0`).

It is unspecified whether or under which conditions the same or different
`Tracer` instances are returned from this functions.

Implementations of this function MUST be prepared that a user calls the
function only once per name+version combination (i.e. caches and re-uses the

> returned `Tracer` instance). If configuration changes should be picked up by

Reviewer 4: ```suggestion
returned `Tracer` instance). If configuration changes SHOULD be picked up by
```

Author: This is not a requirement. Probably I should reword that sentence. What do you think about:

```suggestion
returned `Tracer` instance). If the implementation wants configuration changes to be picked up by
```

users, these changes SHOULD be reflected in already returned objects, not only
if the user calls the API to get a `Tracer` again.

Note: This could, for example, be implemented by storing any mutable
configuration in the `TracerProvider` and having `Tracer` implementation objects
have a reference to the `TracerProvider` from which they were obtained.
If configuration must be stored per-tracer (such as disabling a certain tracer),
the tracer could, for example, do a look-up with its name+version in a map in
the `TracerProvider`, or the `TracerProvider` could maintain a registry of all
returned `Tracer`s and actively update their configuration if it changes.

### Runtimes with multiple deployments/applications <a name="multi-tracerprovider"></a>

Some applications may have to use multiple `TracerProvider` instances, e.g.

> to have different settings (e.g. `SpanProcessor`s) to each (and

Reviewer: ```suggestion
to have different settings (like `SpanProcessor`s) for each (and
```

consequently to the `Tracer`s obtained from them).

For example, runtimes that support multiple deployments or applications might need to
provide a different `TracerProvider` instance to each deployment. To support this,
a global `TracerProvider` registry may delegate calls to create new instances of
`TracerProvider` to a separate `Provider` component, and the runtime may include
its own `Provider` implementation which returns a different `TracerProvider` for
each deployment.

`Provider` instances are registered with the API via some language-specific
mechanism, for instance the `ServiceLoader` class in Java.

## Tracer

The tracer is responsible for creating `Span`s.

Note: Apart from the library name and version
provided to the `TracerProvider` to get them,

> `Tracer`s are not usually expected to not have mutable state on their own

Reviewer: ```suggestion
`Tracer`s are not usually expected to not have mutable state of their own
```

(but they may contain a reference to shared mutable configuration that is

> expected to be owned by the TracerProvider).

Reviewer: ```suggestion
expected to be owned by the `TracerProvider`).
```

### Tracer operations

The `Tracer` MUST provide functions to:

- [Create a new `Span`](#span-creation) (see the section on `Span`)

The `Tracer` SHOULD provide methods to:

- Get the currently active `Span`
- Make a given `Span` as active

The `Tracer` MUST use the [`Context`](../context/context.md)
in order to get and set the current `Span` state and to decide how `Span`s
are passed across process boundaries.

> When getting the current `Span` while there is none that is currently active,

Reviewer 4: It is not clear to me what exactly is "none". Is this intended for when there is no SDK or when there is nothing in the context?

Author: The previous wording was "no currently active `Span`" instead of "none that is currently active", would that be more clear?

The MUST applies only if there is no span in the context, though as per the next paragraph a tracer MAY always return a placeholder span if it feels like it.

Reviewer 3: I find the original wording easier to read. It might be even better if we add sections for getting and setting current span, so that the context of the sentence does not need to be clarified each time.

Author: I changed the wording again in ee86a6a631304cc8f824eaa129734a76d87369a3, because I did not like the old wording with the condition at the end of the sentence. What do you think?

> If there is no currently active `Span`

Reviewer 5: Nit - the line wrapping here seems weird.

Author: I'm prone to use a mixture of semantic line breaking (before the and in this card) and hard line wraps. 😅

Author: In that case I will leave it this way, as when joining the lines they get too long here.

the `Tracer` MUST return a placeholder `Span` with an invalid `SpanContext`
and `IsRecording` being false.

For any two `Tracer`s returned from the same `TracerProvider` and queried at the

> time in the same logical execution unit (see [Context](../context/context.md)),

Reviewer: ```suggestion
same time in the same logical execution unit (see [Context](../context/context.md)),
```

> the active Span MUST be the same unless at least one them

Reviewer: ```suggestion
the active Span MUST be the same unless at least one of the `Tracer`s
```

returns the placeholder span (see above).

> Note: This means that implementations are allowed to "disable" a `Tracer`

Author: @yurishkuro I think https://github.com/open-telemetry/opentelemetry-specification/pull/619/files#r430507815 was meant for this note/the sentence before it. So I will answer it here.

> If one were to follow the pattern suggested here, one would end up with orphaned sub-tree in the trace when a normal child span is created from this placeholder span.

I think passing an explicit span as parent is a very unusual use case. Normally you would pass a Context or use the implicit parent, both of which would still contain the parent "above", if you read the sentence below which remarks that disabled tracers should "do nothing if requested to set the current `Span`".

If you have a hypothetical scenario where different instrumentations, some of which have their tracer disabled, some of which don't, interact by passing explicit Span objects around, you could make it work by returning special "semi-transparent placeholder spans" that behave like the placeholder spun BUT when used as parent or injected, they act as a proxy for their parent. Maybe the requirement could be changed so that instead of returning a placeholder span, only a "non-recording wrapper span" may be returned, that does have a valid span context (namely that of its parent).

Reviewer 3: I'll resolve this to reduce clutter, the problem is described in the next comment.

such that it has no active `Span`.
Such `Tracer`s would typically also return placeholder `Span`s when creating a
new `Span` and to do nothing if requested to set the current `Span`.

The `Tracer` SHOULD provide a way to update its active `Span` and MAY provide
convenience functions to manage a `Span`'s lifetime and the scope in which a
`Span` is active. When an active `Span` is made inactive, the previously-active
`Span` SHOULD be made active. A `Span` maybe finished (i.e. have a non-null end

> time) but still active. A `Span` may be active in one logical execution unit

Reviewer: ```suggestion
time) but still be active. A `Span` may be active in one logical execution unit
```

`Tracer` instances are returned from this functions.

Implementations MUST NOT require users to repeatedly obtain a `Tracer` again
with the same name+version to pick up configuration changes.
This can be achieved either by allowing to work with an outdated configuration or
by ensuring that new configuration applies also to previously returned `Tracer`s.

Note: This could, for example, be implemented by storing any mutable
configuration in the `TracerProvider` and having `Tracer` implementation objects
have a reference to the `TracerProvider` from which they were obtained.
If configuration must be stored per-tracer (such as disabling a certain tracer),
the tracer could, for example, do a look-up with its name+version in a map in
the `TracerProvider`, or the `TracerProvider` could maintain a registry of all
returned `Tracer`s and actively update their configuration if it changes.

## Tracer

The tracer is responsible for creating `Span`s.

Note: Apart from the library name and version
provided to the `TracerProvider` to get them,
`Tracer`s will usually be stateless.
They may contain a reference to shared mutable configuration that is

> expected to be owned by the `TracerProvider`, however.

Author: I think having something like this in the spec is very important, as there was often confusion whether tracers should share SpanProcessors or each should get its own for example. But maybe I can find a way to express this better with less hand-waving.

> expected to be owned by the `TracerProvider`, however.

Author: Does c0298e24e774c2d2ed430fe2b83cf1f71a4db318 address your concerns here?

> expected to be owned by the `TracerProvider`, however.

Reviewer 3: Better. Still irrelevant to the API, could be moved to SDK spec.

> expected to be owned by the `TracerProvider`, however.

Author: I think we should also leave something here, because I think this should not only apply to the official SDK but also be honored by all vendors.

But moving/copying some of this to the SDK spec seems sensible.

> expected to be owned by the `TracerProvider`, however.

Author: I did some more cleanup in 2dbc718c093604aa1032dcf0753f1eaffafbbfe4. I also created #634 because there is no mention in the spec that the Tracer(provider) should call a sampler.

### Tracer operations

The `Tracer` MUST provide functions to:

- [Create a new `Span`](#span-creation) (see the section on `Span`)

The `Tracer` SHOULD provide methods to:

- Get the currently active `Span`
- Mark a given `Span` as active

The `Tracer` MUST use the [`Context`](../context/context.md)
in order to get and set the current `Span` and to decide how `Span`s
are passed across process boundaries.

If there is no currently active `Span`
and the `Tracer` is asked for the current `Span`,
it MUST return a placeholder `Span` with an invalid `SpanContext`
and `IsRecording` being false.

For any two `Tracer`s returned from the same `TracerProvider` and queried at the
same time in the same logical execution unit (see [Context](../context/context.md)),
the active Span MUST be the same unless at least one of the `Tracer`s
returns the placeholder span (see above).
Note: This means that implementations are allowed to "disable" a `Tracer`
such that it has no active `Span`.
Such `Tracer`s would typically also return placeholder `Span`s when creating a

> new `Span` and to do nothing if requested to set the current `Span`.

Reviewer 3: Similar to comment above, I think this description is incorrect. I don't see why disabling a tracer has any impact whatsoever on the current span - the tracer doesn't even KNOW who created the span that the tracer is being asked to make current. It's just a passthrough to `context.get(Span.Key)` / `context.set(Span.Key, span)`.

Regarding startSpan() returning a dummy span when tracer is disabled, this also doesn't make sense, because it leads to broken traces in the following example:
```
func f1(ctx) {
    parentSpan := tracer1.currentSpan(ctx)
    span1 := tracer1.startSpan("one", parent=parentSpan)
    ctx1 := tracer1.setCurrent(ctx, span1)
    f2(ctx1)
}

func f2(ctx) {
    parentSpan := tracer2.currentSpan(ctx)
    span2 := tracer2.startSpan("two", parent=parentSpan)
    ctx2 := tracer2.setCurrent(ctx, span2)
    doSomething(ctx2)
}
```

If `tracer1` is disabled, then per suggestion `span1` is dummy. It's still usable, but no-op. But the problem is that when it's used as parent for `span2`, there is no trace ID / parent ID to use for `span2`.

The way to make this work is to make sure that even though span1 itself is dummy (write operations are no-op), it still has valid span context equal to parentSpan's context. This way span2 will be created as a direct child of the top-level span, and span1 "won't exist".

Author: > The way to make this work is to make sure that even though span1 itself is dummy (write operations are no-op), it still has valid span context equal to parentSpan's context. 

That's a very good idea and also what I wrote in https://github.com/open-telemetry/opentelemetry-specification/pull/619#discussion_r430566671

Author: Should be resolved with 60b19a767fd64bcac3d12b1e40ea7868b407e09b

The `Tracer` SHOULD provide a way to update its active `Span` and MAY provide

> convenience functions to manage a `Span`'s lifetime and the scope in which a

Author: Yes, this is e.g. Java's `withSpan` function. But this seems redundant and better placed in the context management system, so I'll remove this paragraph.

> convenience functions to manage a `Span`'s lifetime and the scope in which a

Author: Removed in 60b19a767fd64bcac3d12b1e40ea7868b407e09b

### Span Creation

A Span MUST ONLY be created via a [`Tracer`](#tracer).

When creating a new `Span`, the `Tracer` MUST allow the caller to specify the
new `Span`'s parent in the form of a `Span` or `SpanContext`. The `Tracer`
SHOULD create each new `Span` as a child of its active `Span`, unless an
explicit parent is provided or the option to create a span without a parent is

> selected, or the currently active `Span` is invalid.

Reviewer 3: ```suggestion
selected.
```

Author: Why this change? If we change this, we must specify what happens when the currently active span is "invalid" (what does that even mean?)

Author: Actually, the fact that I don't know what an "invalid" span should be makes me agree with your suggestion after all. But maybe someone has more input on this, so I'll leave it unresolved for now.

Reviewer 3: precisely, "invalid span" is undefined. Neither is "invalid span context", for that matter. And the whole term "span context" is never introduced before it's mentioned in that sentence. This is just bad juju.

### Span Creation

> A Span MUST ONLY be created via a [`Tracer`](#tracer).

Reviewer 3: The API should not have any other means of creating a span, which would make this line redundant. Which is why I created https://github.com/open-telemetry/opentelemetry-java/issues/1135.

Author: > "The API should not have any other means of creating a span,"

That is exactly what this line means to say. Or to put it another way: That's only true because this line is there.

Author: Hmm, maybe it would be more clear as

```suggestion
There MUST NOT be any API for creating a `Span` other than with a [`Tracer`](#tracer).
```

Author: @yurishkuro What do you think?

Reviewer 3: +1

When creating a new `Span`, the `Tracer` MUST allow the caller to specify the
new `Span`'s parent in the form of a `Span` or `SpanContext`. The `Tracer`

> SHOULD create each new `Span` as a child of its active `Span` unless an

Reviewer 2: ```suggestion
SHOULD create each new `Span` as a child of its active `Span`, unless an
```

explicit parent is provided or the option to create a span without a parent is

> selected, or the current active `Span` is invalid.

Reviewer 2: ```suggestion
selected, or the currently active `Span` is invalid.
```
it's called current_ly_ active everywhere above
